Datetime + duration in locale format

DFreezerR opened the pull request:
Made every date, time, duration, date/time ranges display in current locale format.
Current locale is based on browser lang (I think)
Implemented using `Intl` namespace classes. If class is not found or is not supported in browser then fallback to more generic functions (e.g. `toLocaleString`)

Also added several lang strings into MAL legacy code to translate info titles. 
Other than Chrome, also tested on Firefox since it has some `Intl` stuff missing.

DFreezerR commented:
Need tests fixed

lolamtisch commented on src/_provider/AniList/metaOverview.ts:
```diff
@@ -468,7 +484,7 @@ export class MetaOverview extends MetaOverviewAbstract {
         body: {
           people: i.rating,
-          date: timestampToShortDate(i.createdAt * 1000),
+          date: getDateInLocale(i.createdAt * 1000, 'short'),
```
Wrong in this case. Look at https://github.com/MALSync/MALSync/pull/2698/files?diff=split&w=1#diff-38e47b4eeea96b1634986e199da0917411a0e085aeab0a6949875cd6e7b7bee7R246

DFreezerR replied:
My bad, forgot there should be "relative" date and not "static"

lolamtisch commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+    (Object.keys(duration).length === 1 && (duration.minutes || duration.seconds));
 
-  if (el.m === 1) {
-    output += ` ${el.m} ${api.storage.lang('bookmarksItem_min')}`;
-  } else if (el.m > 1) {
-    output += ` ${el.m} ${api.storage.lang('bookmarksItem_mins')}`;
-  }
+  const inputDuration = minsSecondsOnly
+    ? minsSecsToHoursMins(duration.minutes, duration.seconds)
+    : duration;
```
Should not be handled here. Should be a task for getProgressDateTimeInLocale add different templates for shortTime time if necesseray

DFreezerR replied:
But the progress are only accepts timestamp so no problem with mins. If I'm moving it, how about remove it from here and just call this `minsSecsToHoursMins` function inside meta files to convert to proper hh:mm?

lolamtisch replied:
It adds unneccessary logic to the Intl.DurationFormat wrapper. I see the handling of the time as a 3 steps process.

1. transformer -> Transforms timestamps to durationFormat -> this would just complely remove the need for minsSecsToHoursMins as you can just pass it as a seconds timestamp.
2. formater -> example.: shortTime -> Which takes a durationFormat and takes the hour etc we want to display.
3. Renderer -> Intl.DurationFormat wrapper

Combining 1. and 2. is fine by my book. But 3. should be seperate, as it results in unnecessary amount of different variantations.

lolamtisch replied:
Alternative we could think about going for OOP. Then it would make it simpler to handle the different inputs and output formats. But no idea
`obj.setTimestamp(timestamp)->format('progress')`
`obj.setDate(timestamp)->getDateInLocale()`

DFreezerR replied:
The thing is - `getDurationInLocale` is different from other functions since it just takes `durationFormat` and parses it to string. And I cant just pass it ms units because in the end it will be like this "Duration: 4500000 ms". I can add "transformer" function inside - `timestampToTime` but it converts timestamp relatively to current time. I guess I can do something like this -> Pass (ms duration of anime + `Date.now`) to `timestampToTime`, it calculates how much "time will pass" and returns `durationFormat`, but doesn't it feel like we using function out of purpose?

And yes I agree with your 3 points. As for turning it into OOP idk really, I'll sit for a bit and try something out.

lolamtisch commented on assets/_locales/en/messages.json:
```diff
@@ -458,7 +458,7 @@
     },
     "prediction_Last_anime": {
-        "message": "Last episode released $time$ ago",
+        "message": "Last episode released $time$",
```
This is not a good idea

DFreezerR replied:
Will revert

lolamtisch commented on assets/_locales/en/messages.json:
```diff
@@ -478,7 +478,7 @@
     },
     "prediction_Last_manga": {
-        "message": "Last chapter released $time$ ago",
+        "message": "Last chapter released $time$",
```
This is not a good idea

DFreezerR replied:
Will revert (2)

lolamtisch reviewed (changes requested):
Many things hope it is somewhat understandable. I lost oveview a bit too.

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+// TODO: Delete @ts-expect-error comments after TS adds support for Intl.DurationFormat
+export function getProgressDateTimeInLocale(
+  timestamp: number | null,
+  locale: Intl.LocalesArgument = api.storage.lang('locale'),
```
You are right, I wrote it even before we made locale part of `messages.json`, tried to make functions as much 'generic' as possible. Will remove this unnecessary param

DFreezerR commented on src/utils/time.ts:
```diff
@@ -8,27 +8,66 @@ export interface timeElement {
+  second: 1000,
+} as const;
+
+type durationFormatStyle = 'long' | 'short' | 'narrow' | 'digital';
```
Because there no such type in current ts, so instead of giving error I replaced it with manual type.

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+  locale: Intl.LocalesArgument = api.storage.lang('locale'),
+  style: Intl.DateTimeFormatOptions | undefined = undefined,
+): string {
+  if (!Intl.DateTimeFormat) {
+    return new Date(date).toLocaleString(locale, style);
+  }
```
Oh yes, apparently DateTimeFormat and toLocaleString with locale and style was implemented on the same day. Removing it

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+    (Object.keys(duration).length === 1 && (duration.minutes || duration.seconds));
 
-  if (el.m === 1) {
-    output += ` ${el.m} ${api.storage.lang('bookmarksItem_min')}`;
-  } else if (el.m > 1) {
-    output += ` ${el.m} ${api.storage.lang('bookmarksItem_mins')}`;
-  }
+  const inputDuration = minsSecondsOnly
+    ? minsSecsToHoursMins(duration.minutes, duration.seconds)
+    : duration;
```
Ok, will separate them

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
-    output += ` ${el.h} ${api.storage.lang('bookmarksItem_Hours')}`;
-  }
+// TODO: Delete @ts-expect-error comments after TS adds support for Intl.DurationFormat
+export function getDurationInLocale(
```
You mean move only 2 functions to separate file? Can I inquire why?

DFreezerR commented on src/utils/time.ts:
```diff
@@ -8,27 +8,66 @@ export interface timeElement {
   s: number;
 }
```
Ok, I'll try to replace `timeElement` to `durationFormat` everywhere in `timeToString` `shortTime` etc then

DFreezerR commented on src/utils/time.ts:
```diff
@@ -8,27 +8,66 @@ export interface timeElement {
-    s: seconds,
-  };
+// TODO - Remove when TS adds support for Intl.DurationFormat
+export interface durationFormat {
```
Same as other DurationFormat type, there's none in current ts version

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+}
 
-  return output.trim();
+export function minsSecsToHoursMins(minutes: number = 0, seconds: number = 0): durationFormat {
```
Just thought about it. I believe there's no such provider that gives seconds as duration right? Then I guess I overengineered here, will just convert min to hours+mins in meta

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+    return api.storage.lang('bookmarksItem_now');
   }
+  // @ts-expect-error surely it works
+  if (!Intl.DurationFormat) {
+    const timeString = timeToString(short);
+    return result.isFuture ? timeString : api.storage.lang('bookmarksItem_ago', [timeString]);
+  }
```
Will try

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
+}
 
-  return short;
+export function getDateTimeRangeInLocale(
```
Since I'm already making separate file, let's move it to there. Then as you suggested how about `IntlWrapper.ts`?

DFreezerR commented on src/utils/time.ts:
```diff
@@ -131,76 +170,185 @@ export function shortTime(el: timeElement): timeElement {
-    future = true;
-    diff = tm - curTime;
+// TODO: Delete @ts-expect-error comments after TS adds support for Intl.DurationFormat
+export function getProgressDateTimeInLocale(
```
>Better decide for 1 format.

Right, as I answered somewhere below, I'll rewrite functions for `durationFormat` 

> I think return result.isFuture ? timeString : api.storage.lang('bookmarksItem_ago', [timeString]); should be handled inside of progress.

And this I'll also do.

DFreezerR commented:
Rewritten to separate `IntlWrapper.ts` class. How you feel about it? Tests will do when we agree on final implementation so to not waste time on them.

lolamtisch commented:
@DFreezerR 

I think it makes more sense to handle the duration in a seperate class. I throw together this structure. What do you think of it? This not a blueprint that you need to implement. Just somethig i throw in for discussion.


This is pseudocode. Types and implementations are hacked together and should only give an idea what happens.
```ts
export class IntlDuration {
  constructor(locale = api.storage.lang("locale")) {
    this.locale = locale;
    return this;
  }
  setRelativeTime(relativeTime: number, convertFrom?: keyof typeof dateUnitToMs) {
    this.relativeTime = relativeTime;
    return this;
  }
  setDuration(duration: durationFormat) {
    this.duration = duration;
    return this;
  }
  getDuration() {
    if (!this.duration && !this.relativeTime) throw new Error('No duration or relative time set');
    if (!this.duration) this.format();
    return this.duration;
  }

  format(DurationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal') {
    if (!this.relativeTime) throw new Error('No duration or relative time set');
    switch (DurationStyle) {
      // I just wanted to make sure this is possible. It does not need to be implemented, as I think not needed anywhere.
      case 'm/h/d/y':
        this.duration = durationOnlyWithMinutesHoursDaysYears(this.relativeTime); 
        break;
      case 'normal':
        this.duration = msToDuration(this.relativeTime);
        break;
      case 'progress':
        tempDuration = msToDuration(this.relativeTime);
        this.duration = shortTime(tempDuration);
        break;
      default:
        throw new Error('Unknown DurationStyle');
    }
  }

  text(style: durationFormatStyle = 'narrow') {
    if (fallback) {}
    return new Intl.DurationFormat(this.locale, { style }).format(this.getDuration());
  }
}

// intl.setDuration({ hours, minutes }).Duration.get() -> new IntlDuration().setDuration({ hours, minutes }).text()


export class IntlWrapper {
  protected mainDate = Date.now();
  protected comparisionDate = Date.now();

  constructor(locale = api.storage.lang("locale")) {
    this.locale = locale;
    return this;
  }
  setDate(date) {
    this.mainDate = date;
  }
  setComparisionDate(date) {
    this.comparisionDate = date;
  }
  setTimestamp(timestamp: number, convertFrom?: keyof typeof dateUnitToMs) {
    // Convert either to timestamp to date or date to timestamp
    this.mainDate = timestampToDate();
  }
  isNow(){}
  isInFuture(){}
  getRelativeTime(style: durationFormatStyle = 'narrow', DurationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal'){
    relative = Math.abs(this.mainDate - this.comparisionDate);
    durationObj = new IntlDuration(this.locale).setRelativeTime(relative).format(DurationStyle);
    return durationObj.text(style);
  }
  getRelativeTimeText(style: durationFormatStyle = 'narrow', DurationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal'){
    if (this.relative.isNow()) return 'now';
    timeString = this.relative.get(style, DurationStyle);
    return this.relative.isFuture() ? timeString: timeString + ' ago';
  }
  DateTime(style?: Intl.DateTimeFormatOptions){}
  Range(style?: Intl.DateTimeFormatOptions) {
    from = this.mainDate;
    to = this.comparisionDate;
  }
}
```


I think if you want to give along some preset. I think using a factory function is a lot clearer to what is happening.

```ts
export class IntlWrapper {
  durationFormatStyle: durationFormatStyle = 'narrow';
  durationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal';
  DateTimeFormatOptions: Intl.DateTimeFormatOptions = {};
  .
  .
  .
  // Example
  getRelativeTime(){
    relative = Math.abs(this.mainDate - this.comparisionDate);
    durationObj = new IntlDuration(this.locale).setRelativeTime(relative).format(this.durationDurationStyle);
    return durationObj.text(this.durationFormatStyle);
  }
  .
  .
  .
  static create(rule: "none" | "date" | "time" | "progress" = "none", locale = api.storage.lang("locale")) {
    obj = new IntlWrapper(locale);
    
    switch (rule) {
      case "progress":
        obj.setDurationFormatStyle("narrow");
        obj.setDurationStyle("progress");
        break;
      case "date":
        obj.setDateTimeFormatOptions({ dateStyle: "medium" });
        break;
      default:
        break;
    }
    return obj;
  }
}


```

DFreezerR commented:
So your points:
1. move duration + progress into separate class and just call them from `IntlWrapper`? I got that 👍
2. create factory function inside `IntlWrapper` to increase readability? Got questions

For 2nd point I wanted to ask questions:
1. You wrote 2 examples in both code blocks, which one I should consider?
> `// intl.setDuration({ hours, minutes }).Duration.get() -> new IntlDuration().setDuration({ hours, minutes }).text()`

or
> `static create(rule: "none" | "date" | "time" | "progress" = "none", locale = api.storage.lang("locale"))`

Also I wanted to make code more "generic" so user can specify (or not) styles by themselves, but with `create` function it is not possible, but I guess more parameters may be added to it.
With `create` approach I can't completely imagine how code will be looked when getting value, correct me if I'm wrong. Something like this? - 
`new IntlWrapper().setTimestamp(this.getPredictionTimestamp()).Progress.get()` 
will turn into 
`IntlWrapper.create('progress', 'digital').getRelativeTimeText(this.getPredictionTimestamp())`

2. Is that correct that you mean that `create` function only for progress and duration since they are from separate class? 
Because as you saw I tried making getting every field with the same "schema", like you want date then you set timestamp or date and then call DateTime->get(). With `create` it will be that duration and progress have separate "method of getting value" hope you get what I mean.

DFreezerR commented:
So I thought a bit and based on what you said I'm seeing it like this:
1. Use `create` to define what type of data you want + optional args like style and locale:
```ts
//user call
 IntlWrapper.create('datetime', 'short')
//actual function 
 IntlWrapper.create(rule, style, locale)
 ```
2. Then it "prepares" the ground, setting required fields like styles:
```ts
//inside create function
this.setStyle('short')
this.setLocale(locale)
this.setRule('datetime')
...
```
3. Use function to for needed data (e.g getDateTime) to set up input data like timestamp:
```ts
//user call
  IntlWrapper.create('datetime', 'short').getDateTime(timestamp)
//or
  IntlWrapper.create('duration', 'narrow').getRelativeTime(timestamp)
```
4. Then it will call specified transform + format functions for selected type of data
```ts
//inside a function getDateTime
const time = this.setTimestamp(timestamp).transform(rule).format(rule)
return time.text(style); //returning in text form

//if this is progress or duration (when calling getRelativeTime for example)
const duration = new IntlDuration().setTimestamp(timestamp).transform(rule).format(rule)
return time.text(style); //returning in text form
```
5. Based on input rule it will determine how to transform and format data (like I did in last commit)
```ts
//inside transform/format
this.time = shortTime(this.timestamp)
//or
this.time = timeToTimestamp(this.duration)
...
//other manipulations
```
Tell if I'm being somewhere wrong. All above it just an example, there obviously will be more fields/variables.

lolamtisch commented:
You misunderstood the create function. It is a factory function. Meaning in this case it is like a optional preset. 
You can still set the different styles by hand. 
`new IntlWrapper().setDurantionStyle('wide').setOtherStyle(xyz).getRelativeText()`
Is still possible or you use the `IntlWrapper.create('progress').getRelativeText()` to stay consistent.

I am on my phone I will look more in detail into your answer tomorrow.

DFreezerR commented:
>  it is like a optional preset. You can still set the different styles by hand.

Ah it's optional presets, I got it now.

DFreezerR commented:
UPD:

Should we really call IntlDuration from IntlWrapper? I just tried and it's kinda ugly. I guess the main thing I want to ask - what relationship between these 2 classes should be?
1. Completely separate - if user wants duration/progress he will create `IntlDuration` instance directly (either with "new" or `create()`). Both classes will have their own preset create function with their own rules. And you cant call `IntlDuration` from `IntlWrapper`
2. Completely separate with possibility to get `IntlDuration` from `IntlWrapper` - everything is the same but user can ALSO get duration/progress from `IntlWrapper`, either with `create()->getDuration(input)` or with `new IntlWrapper()->setTimestamp(input)->getDuration(); The `getDuration` function will just create `IntlDuration` instance, process stuff and return text. That looks bad for me because in that case there will be a ton of duplicate code in `IntlWrapper` like duration fields, setters for styles etc.
3. IntlDuration can be accessed only from IntlWrapper - there won't be `create` preset function in `IntlDuration`.  Class won't be exported. To get duration/progress you have to use the same methods from 2p. (from constructor or `create`). There won't be duplicate code in `IntlWrapper` since `IntlDuration` class will have only transform, format and `toText` functions without having it's own fields for duration style and setters (they will come as parameters from `IntlWrapper`)

Thoughts?

lolamtisch commented:
@DFreezerR 
How I see it, is that they are in different domains. IntlDuration runs on relative time and IntlWrapper on Dates/timestamps. We could change the name to refrect that better, but no idea. But as they work in different domains it makes sense that they are seperate utility classes. That is why I do not see any problem in IntlWrapper calling IntlDuration, as it is a utility and I see no problem that IntlDuration gets directly used elsewhere, where only relative times are of interesse. 

But I did not see a problem of IntlWrapper calling IntlDuration to begin with, did I miss something?


For IntlDuration I think I would not use a factory, as we only have one option, so it could be added directly to the constructor

```
export class IntlDuration {
  constructor(DurationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal', locale = api.storage.lang("locale")) {
```

DFreezerR commented:
>  We could change the name to refrect that better, but no idea. 

Yeah, that's what I thought.

> But I did not see a problem of IntlWrapper calling IntlDuration to begin with, did I miss something?

Well, the thing I mean is - for each data we provide with class, we can get its preset with factory function `create` (datatime, time, date, range) and we have specified function to get this date (getDateTime, getRange etc.) but for duration/progress we don't have  it's preset in `create` function but still have "get" function. I just thought it might not be consistent an not "perfect". Should I just forget it and it's all good?

lolamtisch commented:
In that case I think it makes more sense to drop the create function/presets in general and instead focus more in creating utility functions. As progress is handled centrailised already we do not really need an preset and if needed for the other things we can just create an function in the other time.ts utility.

Not sure if IntlRange can really be handled like that.

```TS
class IntlDuration {
    constructor(locale = api.storage.lang("locale")) {
        this.locale = locale;
        return this;
    }
    setRelativeTime(relativeTime: number, convertFrom?: keyof typeof dateUnitToMs, style: 'normal' | 'progress' | 'm/h/d/y' = 'normal') {
        this.duration = this.relativeTimeToDuration(relativeTime)
        return this;
    }
    relativeTimeToDuration(relativeTime, style: 'normal' | 'progress' | 'm/h/d/y' = 'normal') {
        if (!relativeTime) throw new Error('No duration or relative time set');
        switch (this.style) {
            // I just wanted to make sure this is possible. It does not need to be implemented, as I think not needed anywhere.
            case 'm/h/d/y':
                this.duration = durationOnlyWithMinutesHoursDaysYears(relativeTime);
                break;
            case 'normal':
                this.duration = msToDuration(relativeTime);
                break;
            case 'progress':
                tempDuration = msToDuration(relativeTime);
                this.duration = shortTime(tempDuration);
                break;
            default:
                throw new Error('Unknown DurationStyle');
        }
    }
    setDuration(duration: durationFormat) {
        this.duration = duration;
        return this;
    }
    getDuration() {
        return this.duration;
    }
    text(style: durationFormatStyle = 'narrow') {
        if (fallback) { }
        return new Intl.DurationFormat(this.locale, { style }).format(this.getDuration());
    }
}

export class IntlDateTime {
    protected mainDate = Date.now();

    setDate(date) {
        this.mainDate = date;
    }
    
    setTimestamp(timestamp: number, convertFrom?: keyof typeof dateUnitToMs) {
        // Convert either to timestamp to date or date to timestamp
        this.mainDate = timestampToDate();
    }
    isNow(coparision = Date.now()) { }
    isInFuture(coparision = Date.now()) { }
    getdateTimeText(style?: Intl.DateTimeFormatOptions) { }
    getRelativeDuration(durationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal', coparision = Date.now()) {
        relative = Math.abs(this.mainDate - coparision);
        return new IntlDuration(this.locale).setRelativeTime(relative, 'ms', durationStyle);
    }
    getRelativeTime(style: durationFormatStyle = 'narrow', durationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal', coparision = Date.now()) {
        return this.getRelativeDuration(style, durationStyle, coparision).text(style);
    }
    getRelativeTimeText(style: durationFormatStyle = 'narrow', durationStyle: 'normal' | 'progress' | 'm/h/d/y' = 'normal', coparision = Date.now()) {
        if (this.isNow(coparision)) return 'now';
        timeString = this.getRelativeTime(style, DurationStyle, coparision);
        return this.isInFuture(coparision) ? timeString : timeString + ' ago';
    }
}

export class IntlRange {
    protected mainDate: IntlDateTime;
    protected comparisionDate: IntlDateTime;

    Range(style?: Intl.DateTimeFormatOptions) {
        return `${this.mainDate.getdateTimeText(style)} - ${this.comparisionDate.getdateTimeText(style)}`;
    }
}
```

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -26,139 +28,111 @@ const dateUnitToMs = {
+    return true;
+  }
+
+  protected transform() {
```
Sure

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
-    return new Intl.DateTimeFormat(this.locale, style);
+  static toTimestamp(time: number, from: keyof typeof dateUnitToMs): number {
+    return Date.now() + time * dateUnitToMs[from];
   }
```
True

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -26,139 +28,111 @@ const dateUnitToMs = {
-      isFuture: false,
-    },
-  };
+  protected relativeTime?: number;
```
That would mean passing a timestamp will look like this - `new IntlDuration({milliseconds: someTimestamp})`
And if we are to convert minutes to hh:mm - `new IntlDuration({minutes: 150}, 'minutes')` + all the timestamp calculations needs to be done with `duration.milliseconds` like `toTimestamp`. Looks a bit limited. You sure?

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -26,139 +28,111 @@ const dateUnitToMs = {
 
-  // Input variables
-  protected from: Date | number = new Date();
+  protected isFuture = false;
```
How it is not? If we move it to `IntlDateTime` then how user will get now/future states if he use `IntlDuration` directly?

lolamtisch replied:
Because that is impossible? There is no now or future in relative time. Because it can be between anything.

DFreezerR replied:
Oooh, you want to make it work between any dates? I only thought about relative to "now"

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
-      if (value <= 30 && key === 'seconds') break;
-    }
-    return true;
+export class IntlDateTime {
```
Same as above

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
-    return true;
+export class IntlDateTime {
+  // Output variables
+  protected text: string = '';
```
Ok, seems logical

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -26,139 +28,111 @@ const dateUnitToMs = {
+  protected isFuture = false;
 
-  protected to: Date | number = new Date();
+  protected isNow = false;
```
I mean isNow only working for "Progress" rule, for events and other stuff you should use different rules no?

DFreezerR replied:
Ok, now I get it, since it's better to have "relative time" relative to specific date and not just "now" we move it to DateTime class since you can specify date here.

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
+  getText(): string {
+    this.transform().format();
+    if (!this.duration) return '';
+    if (this.isNow) return api.storage.lang('bookmarksItem_now');
```
I thought you said somewhere above that these classes can be used directly without each other hence I made them independent?

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
+  }
+
+  // Processing
+  protected format() {
```
Ok

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -26,139 +28,111 @@ const dateUnitToMs = {
-    this.timestamp = this.from.getTime();
+  setRelativeTime(relativeTime: number, convertFrom?: keyof typeof dateUnitToMs) {
+    this.relativeTime = Number(relativeTime);
+    if (convertFrom) this.toTimestamp(convertFrom);
```
So you mean completely working with `durationFormat` object? The only use for `relativeTime` was that the input could be like 150 (mins) and that number easily convertable to timestamp and then to `durationFormat`. If I use only `durationFormat` as an input param I could not convert 150 (in this case `{minutes: 150}`) to `{hours: 2, minutes: 30}` because `timestampToTime` only uses timestamps. I guess I could make another function to convert something like  `{minutes: 150}` to `{hours: 2, minutes: 30}`?

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
+    return this.text;
+  }
+
+  getRange(from: Date | number | string, to: Date | number | string) {
+    const range = new IntlRange(from, to, this.dateTimeFormatStyle, this.locale);
+    return range.getText();
+  }
```
Ok

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
+
+  constructor(
+    date: Date | number | string = new Date(),
+    style: Intl.DateTimeFormatOptions = { dateStyle: 'medium' },
```
Removed, made style as optional param in "get" functions

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
+  protected dateTimeFormatStyle: Intl.DateTimeFormatOptions;
+
+  constructor(
+    date: Date | number | string = new Date(),
```
Will try

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -167,119 +141,207 @@ export class IntlWrapper {
-      },
-    },
-  };
+  getRelative(
```
Ok, lets me again clarify this, summarizing all above what you mean is:
1. `IntlDuration` should be created from `durationFormat` object ONLY
2. If we call it from `IntlDateTime` using Date object we pass it like this - `new IntlDuration({milliseconds: this.date.getTime()})`
3. Dont handle now/future in `IntlDuration` but create 2 functions in `IntlDateTime` 1st could return "Now" and isFuture state and 2nd could not (Maybe just do 1 function but with param?)

lolamtisch replied:
@DFreezerR 

1 + 2. I see the input of IntlDuration was duration + timestamp. I was thinking it was duration + relative time (because of the setRelativeTime function). But I see the problem you have.
We have 3 types of input.
- {minutes: 150} -> 2h 30min // apply an format
- {hours: 2, minutes: 30} -> 2h 30min // output the duration object like passed
- 9.000.000 -> 2h 30min (3. is not 100% necessary but as it it makes sense to handle it like that internaly why not expose it) // ms relative time

I think in this case it would be the best to remove the input from the constructor and instead have a set function for it. So we have multiple entrypoints.

```ts
(new IntlDuration()).setFormatedDuration({{hours: 1, minutes: 90}, 'M/H/D/Y'); -> 2h 30min
(new IntlDuration()).setDuration({hours: 2, minutes: 30}); -> 2h 30min
(new IntlDuration()).setRelativeTime(9.000.000, 'ms',  'M/H/D/Y' ); -> 2h 30min
```
```ts
  setDurationFormated(duration, style: 'normal' | 'progress' | 'm/h/d/y' = 'normal') {
    const relativeTime = durationToMs(duration);
    this.setRelativeTime(relativeTime, 'ms', style);
    return this;
  }

  setRelativeTime(relativeTime: number, convertFrom?: keyof typeof dateUnitToMs, style: 'normal' | 'progress' | 'm/h/d/y' = 'normal') {
    const duration = this.relativeTimeToDuration(relativeTime, style); // Transforms and applies the format
    this.setDuration(duration);
    return this;
  }

  setDuration(duration) {
    this.duration = duration;
    return this;
  }
```

4.
Yes it should be compared to current date.

3.
Return an boolean for both.

DFreezerR replied:
> I think in this case it would be the best to remove the input from the constructor and instead have a set function for it.

Yeah, I think I had something like this in the beginning
Oh well, will make adjustments , since I now understood what you tried to say

Hope you don't mind our "slow walking" with this PR

DFreezerR commented:
- Added different entry points to `IntlDuration`
- Moved now/future logic to `IntlDateTime`
- Removed internal styles, moved them as processing function param
- Simplified classes for better readability
- Removed `IntlRange` call from `IntlDateTime`
- Added functions to `IntlDateTime` to choose from returning duration with "now" or not
- Added `checkForNow`

lolamtisch commented on src/utils/IntlWrapper.ts:
```diff
@@ -66,69 +46,47 @@ export class IntlDuration {
+    style: durationStyle = 'Duration',
+    relativeTo: Date | number = new Date(),
+  ) {
+    const duration = IntlDuration.timestampToDuration(timestamp, relativeTo);
```
But I would just call setRelativeTime directly with the relative time, as it is just a simple absolute substraction then. Currently there is an unecessary transformation into duration. 

Because Currently it is `date` -> `relativetTime` -> `duration` -> `relativeTime` -> `duration`

lolamtisch commented on src/utils/IntlWrapper.ts:
```diff
@@ -319,28 +220,27 @@ export class IntlDateTime {
+    const duration = relative.getDuration();
+    if (!duration) return '';
+    return relative.getRelativeText(format);
   }
```
Not what I was expecting of the function. Was thinking of something more in this direction. Specially as yours has a lot of duplicate code and does not return what I was expecting

```ts
    getRelativeNowText(style: durationStyle = 'Duration', format?: durationFormatStyle) {
      if (!this.isValidDate()) return '';
      const relative = new IntlDuration(this.locale);
      relative.setTimestamp(this.date.getTime(), style);
      return relative.getRelativeText(format);
    }
    getRelativeTimeText(style: durationStyle = 'Duration', format?: durationFormatStyle) {
        if (this.isNow()) return api.storage.lang('bookmarksItem_now');
        timeString = this.getRelativeNowText(style, format);
        return this.isFuture() ? timeString :  api.storage.lang('bookmarksItem_agoxyz',[timeString ]);
    }
```

DFreezerR replied:
Oh I see, thanks for pinpointing this
Btw we already have `prediction_Episode_****` and `prediction_Last_****` in `progress.ts` which returns strings with "ago", do we even need them? I guess I just delete those lines in `progress.ts` to avoid double "ago" and remove `getLastText` and `getPredictionText` since we already returning final text from `IntlDatetime`

lolamtisch replied:
As it was not happening in progress at all. I though you wanted to first do the structure and I did not point it out. But probably because of our bad progress structure. getLast is what is used nearly anywhere and can directly use FriendlyText. getLastText is used very seldom, was thinking of removing it, that is why it is ok to just implement it again there, as we do not really care. 

I would not handle the `getLastText` inside of the getRelativeNowFriendlyText, as it is not that general and we do not really care.

```ts
// progress.ts
  getLast(): string {
    const timestamp = Number(this.getLastTimestamp());
    return new IntlDateTime(timestamp).getRelativeNowFriendlyText('Progress'); // returns 2d ago
  }
  getLastText(): string {
    const dt = new IntlDateTime(Number(timestamp));
    if (dt.isNow()) return api.storage.lang('bookmarksItem_now');
    return dt.isFuture() ? timeString :  api.storage.lang(`prediction_Last_${this.type}`, [timeString ]); // returns Last episode was 2d ago
  }
```

DFreezerR replied:
Done

lolamtisch reviewed (changes requested):
> Hope you don't mind our "slow walking" with this PR

I am fine with it. I was more concerned that I am annoying you too much with all the change requests. But I think we are quite close now.

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -66,69 +46,47 @@ export class IntlDuration {
+  // For {hours: 2, minutes: 15}
+  setDuration(duration: durationFormat, style: durationStyle = 'Duration') {
     this.duration = duration;
+    this.process(style);
```
Yes, make sense. Style param also not needed in that case.

DFreezerR commented on src/utils/IntlWrapper.ts:
```diff
@@ -66,69 +46,47 @@ export class IntlDuration {
-    switch (this.durationStyle) {
-      case 'Duration':
-        break;
+  protected process(style: durationStyle) {
```
How about adding additional param to `relativeToDuration` for `convertFrom` so it will look like `IntlDuration.relativeToDuration(WhatToConvert, WhichToConvertFrom, WhatToConvertInto)` or we rely on the fact that user will specify unit to convert from as the first one like `IntlDuration.relativeToDuration(9000, ['seconds', 'minutes', 'hours', 'days', 'years']);` - 'seconds' is input here

lolamtisch replied:
To keep everything as simple as possible I would use ms as early as possible and have a different function just for the timeIntervalToMs

DFreezerR replied:
Done it

DFreezerR commented:
- Basically what commit says

lolamtisch commented on src/utils/progress.ts:
```diff
@@ -103,27 +103,42 @@ export class Progress {
+    const timestamp = Number(this.getLastTimestamp());
+    if (Number.isNaN(timestamp)) return '';
+    const dt = new IntlDateTime(Number(timestamp));
+    const time = dt.getRelativeNowText('Progress');
```
I think we could use the long version here. What do you think?

DFreezerR replied:
Seem logical to me. That string is used only in progress pill hover, so no need to save space

lolamtisch commented on src/utils/progress.ts:
```diff
@@ -103,27 +103,42 @@ export class Progress {
+    const timestamp = Number(this.getPredictionTimestamp());
+    if (Number.isNaN(timestamp)) return '';
+    const dt = new IntlDateTime(Number(timestamp));
+    const time = dt.getRelativeNowText('Progress');
```
I think we could use the long version here. What do you think?

lolamtisch reviewed (changes requested):
I think I am happy with the Intl classes now

DFreezerR commented on src/_minimal/components/overview/overview-info.vue:
```diff
@@ -92,11 +92,11 @@ defineProps({
 function getTitle(item) {
   if (item.lastEp && item.lastEp.timestamp) {
-    return timestampToShortTime(item.lastEp.timestamp);
+    return new IntlDateTime(item.lastEp.timestamp).getRelativeNowText('Progress');
```
In my testing that was working, but will add parse to number just in case something change

DFreezerR commented on src/_minimal/components/settings/settings-list-sync.vue:
```diff
@@ -185,7 +195,9 @@
                   <span v-if="slave.diff && slave.diff.finishDate !== undefined">
```
Somehow I missed it, thanks

DFreezerR commented on src/_provider/AniList/metaOverview.ts:
```diff
@@ -468,7 +484,9 @@ export class MetaOverview extends MetaOverviewAbstract {
       reviews.push({
         body: {
           people: i.rating,
-          date: timestampToShortDate(i.createdAt * 1000),
```
Wait, we want relative time on review date? Can't remember. I will look something like this? (just edited html)
![image](https://github.com/user-attachments/assets/b62ef059-a910-4a08-a97a-aa5744c7db82)

lolamtisch replied:
Not sure why this was relative to begin with. The normal date is probably a lot better here

DFreezerR replied:
Same

DFreezerR commented:
- Fixes
- Added missing dates

lolamtisch commented on src/utils/IntlWrapper.ts:
```diff
@@ -0,0 +1,376 @@
+    if (!this.duration) return '';
+    if (this.isFallback) return timeToString(this.duration);
+    // @ts-expect-error surely it works
+    return new Intl.DurationFormat(this.locale, { style: format }).format(this.duration);
```
I am not fully happy with output for the narrow progress outputs. Long does not look good either.

I tested around using narrow but still having the "most significant bit/dateunit" as long. It seemed a good compromise between narrow and long. I looked at english, spanish and german and it looked fine. What do you think? 
![image](https://github.com/user-attachments/assets/b182b3b2-9fb4-490d-83fe-3ab01b5b9345)
![image](https://github.com/user-attachments/assets/09db13f1-8299-4370-ab89-478a35cbf49b)
![image](https://github.com/user-attachments/assets/10873f14-f2ce-4019-b7f4-0c119426bd9f)
![image](https://github.com/user-attachments/assets/d0db42e4-1d01-42ed-ad7e-23d21bd7c3e3)
![image](https://github.com/user-attachments/assets/9cbb0d55-0fd1-4fca-9a09-2bd2898406d1)


Examples:
 `{days: 2, hours: 45}` -> `{ days: 'long', style: 'narrow'}`
`{hours: 2, minutes 45}` -> `{ hours: 'long', style: 'narrow'}`
etc...

`Intl.DurationFormat(this.locale, { days: 'long', style: 'narrow'}).format(this.duration);`

Implementation wise I think the best way is to add an additional custom durationFormatStyle(`narrowMSB/primaryLongOthersNarrow/significantLongNarrow`) and then handle it in getRelativeText.

DFreezerR replied:
I assume you don't like "short"? 
Well I like how your screenshots look like, so why not. Butto be honest I have no clue what `narrowMSB` means and what diff between other 2 lmao
I guess `significantLongNarrow` is the first date unit is "long" format and any other is "narrow"?

lolamtisch reviewed (changes requested):
We are close

DFreezerR commented on src/_minimal/components/overview/overview-info.vue:
```diff
@@ -92,11 +92,11 @@ defineProps({
 function getTitle(item) {
   if (item.lastEp && item.lastEp.timestamp) {
-    return timestampToShortTime(item.lastEp.timestamp);
+    return new IntlDateTime(item.lastEp.timestamp).getRelativeNowText('Progress');
```
Both done

DFreezerR commented:
Check this out.

- Added number string check in IntlDateTime
- Replaced style type to options object for IntlDuration class to match  IntlDateTime format options param
- Added new style for "first unit is long everything is narrow"

lolamtisch commented on src/utils/IntlWrapper.ts:
```diff
@@ -0,0 +1,412 @@
+    if (this.isFallback) return timeToString(this.duration);
+    const newOptions: DurationFormatOptions = options;
+    if (options.style === 'significantLongNarrow') {
+      newOptions[Object.keys(this.duration)[0]] = 'long';
```
It is really easy to get around it with `Object.keys({days: 2, years: 2})[0]`

I think the better idea would be to use dateUnitToMs.find() so that it iterates from years down until it finds one entry.

DFreezerR replied:
oh... didn't thought about that case

DFreezerR replied:
There's no `find` so just checking existence of key

DFreezerR commented:
- Fixed getting most significant date unit

lolamtisch commented:
Oh forgot to answer. From my side it is done. Now only the tests have to fixed/redone

DFreezerR commented:
> Oh forgot to answer

No problem
Tests are here

DFreezerR commented:
And you for your time
